src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs

```diff
@@ -374,6 +380,672 @@ static void appendModifiers(ImmutableArray<CustomModifier> customModifiers, Stri
+                    Cci.CallingConvention.Unmanaged => " unmanaged",
+                    Cci.CallingConvention.CDecl => " unmanaged[CDecl]",
+                    Cci.CallingConvention.Standard => " unmanaged[Stdcall]",
+                    Cci.CallingConvention.ThisCall => " unmanaged[Thiscall]",
+                    Cci.CallingConvention.FastCall => " unmanaged[Fastcall]",
+                    _ => throw ExceptionUtilities.UnexpectedValue(signature.CallingConvention)
+                };
+
+                builder.Append(callingConvention);
+                appendCallingConventionTypes(signature.UnmanagedCallingConventionTypes, builder);
+
+                bool needComma = false;
+                builder.Append('<');
+                ImmutableArray<ParameterSymbol> parameters = signature.Parameters;
+                for (int i = 0; i < parameters.Length; i++)
+                {
+                    if (needComma)
+                    {
+                        builder.Append(", ");
+                    }
+
+                    ParameterSymbol parameter = parameters[i];
+                    appendRefKind(parameter.RefKind, builder, forParameter: true);
+                    appendTypeWithAnnotation(parameter.TypeWithAnnotations, builder);
+                    Debug.Assert(parameter.TypeWithAnnotations.CustomModifiers.IsEmpty);
+                    needComma = true;
+                }
+
+                if (needComma)
+                {
+                    builder.Append(", ");
+                }
+
+                appendRefKind(signature.RefKind, builder);
+                appendTypeWithAnnotation(signature.ReturnTypeWithAnnotations, builder);
+
+                builder.Append('>');
+            }
+
+            static void appendRefKind(RefKind refKind, StringBuilder builder, bool forParameter = false)
+            {
+                builder.Append(refKind switch
+                {
+                    RefKind.None => "",
+                    RefKind.Ref => "ref ",
+                    RefKind.Out => "out ",
+                    RefKind.In when forParameter => "in ",
+                    RefKind.RefReadOnly => "ref readonly ",
+                    RefKind.RefReadOnlyParameter => "ref readonly ",
+                    _ => throw ExceptionUtilities.UnexpectedValue(refKind)
+                });
+            }
+
+            static void appendCallingConventionTypes(ImmutableArray<NamedTypeSymbol> callingConventionTypes, StringBuilder builder)
+            {
+                if (callingConventionTypes.IsEmpty)
+                {
+                    return;
+                }
+
+                builder.Append('[');
+                for (int i = 0; i < callingConventionTypes.Length; i++)
+                {
+                    if (i > 0)
+                    {
+                        builder.Append(", ");
+                    }
+
+                    Debug.Assert(callingConventionTypes[i].Name.StartsWith("CallConv", StringComparison.Ordinal));
```

Comment:
>Debug.Assert(callingConventionTypes[i].Name.StartsWith("CallConv", StringComparison.Ordinal)); [](http://example.com/codeflow?start=20&length=94)

Can we really rely on this in an invalid code?

Reply:
Yes, the logic in `UnmanagedCallingConventionTypes` only includes modifiers that conform to `FunctionPointerTypeSymbol.IsCallingConventionModifier` which includes a check for `modifierType.Name.StartsWith("CallConv", StringComparison.Ordinal)`.